shared/src/main/scala/coop/rchain/dag/DagOps.scala

```diff
@@ -0,0 +1,158 @@
+import cats.implicits._
+import cats.{Applicative, Eval, Monad}
+import coop.rchain.shared.StreamT
+
+import scala.collection.Seq
+import scala.collection.generic.Sorted
+import scala.collection.immutable.{HashSet, Queue}
+
+trait DagReader[F[_], A] {
+  def children(vertex: A): F[Option[Set[A]]]
+  def parents(vertex: A): F[Option[Set[A]]]
+}
+
+trait DagUpdater[F[_], A] {
+  def addEdge(child: A, parent: A): F[InMemDAG[F, A]]
+  def remove(vertex: A): F[Unit]
+}
+
+final case class InMemDAG[F[_]: Applicative, A](
+    childrenMap: Map[A, Set[A]],
+    parentsMap: Map[A, Set[A]]
+) extends DagReader[F, A]
+    with DagUpdater[F, A] {
+  override def children(vertex: A): F[Option[Set[A]]] =
+    childrenMap.get(vertex).pure[F]
+  override def parents(vertex: A): F[Option[Set[A]]] = parentsMap.get(vertex).pure[F]
+  override def addEdge(child: A, parent: A): F[InMemDAG[F, A]] =
+    new InMemDAG[F, A](
+      this.childrenMap.updated(parent, childrenMap.getOrElse(parent, Set.empty[A]) + child),
+      this.parentsMap.updated(child, parentsMap.getOrElse(child, Set.empty[A]) + parent)
+    ).pure[F]
+
+  override def remove(vertex: A): F[Unit] = ???
+}
+
+object DagOps {
+  def bfTraverseF[F[_]: Monad, A](start: List[A])(neighbours: A => F[List[A]]): StreamT[F, A] = {
+    def build(q: Queue[A], prevVisited: HashSet[A]): F[StreamT[F, A]] =
+      if (q.isEmpty) StreamT.empty[F, A].pure[F]
+      else {
+        val (curr, rest) = q.dequeue
+        if (prevVisited(curr)) build(rest, prevVisited)
+        else
+          for {
+            ns      <- neighbours(curr)
+            visited = prevVisited + curr
+            newQ    = rest.enqueue[A](ns.filterNot(visited))
+          } yield StreamT.cons(curr, Eval.always(build(newQ, visited)))
+      }
+
+    StreamT.delay(Eval.now(build(Queue.empty[A].enqueue[A](start), HashSet.empty[A])))
+  }
+
+  def bfTraverse[A](start: List[A])(neighbours: A => List[A]): Stream[A] = {
+    def build(q: Queue[A], prevVisited: HashSet[A]): Stream[A] =
+      if (q.isEmpty) Stream.empty[A]
+      else {
+        val (curr, rest) = q.dequeue
+        if (prevVisited(curr)) build(rest, prevVisited)
+        else {
+          val ns      = neighbours(curr)
+          val visited = prevVisited + curr
+          val newQ    = rest.enqueue[A](ns.filterNot(visited))
+          Stream.cons(curr, build(newQ, visited))
+        }
+      }
+
+    build(Queue.empty[A].enqueue[A](start), HashSet.empty[A])
+  }
+
+  /**
+    * branch = tip block + all descendants through which tip is connected to base block.
+    * @return sequence of blocks inside branch
+    */
+  def computeBranch[F[_]: Monad, A](tip: A, base: A, dag: DagReader[F, A]): F[(A, Seq[A])] =
+    for {
+      // tip + all ancestors down to LFB block height
+      // traverse from tip to LFB through parents, save blocks visited
+      tipAncestors <- (Seq(tip), Seq(tip)).tailRecM[F, Seq[A]] {
+                       case (acc, blocksToProceed) =>
```

Comment:
I just want to say this can be a performance issue if the traverse goes very long. 
The performance issue would be showed like all memory is occupied and the node is doing gc from my perspective.
This is from the experience which I did on the pr which traverse the whole dag and accumulate finalized blocks.
https://github.com/rchain/rchain/pull/3258/files#diff-74cd1658ab1ea29ab3304579f96ff1f6a6f4488b0b85b5cbeedcada721016ce7R121-R132 this is the algorithm which would cause performance issue.
I fixed that by using `Stream` like this https://github.com/rchain/rchain/pull/3263/files#diff-b2346ef03013f08f23c57bb943e5e8e9417d11517e8e448c9aaa9703c7b83ce6R26-R41 .

Below is my guess for the reason.
Because `acc` is accumulating and each iteration would create a new Seq in memory, memory would accumulate in O(n**2) way which can easily consume all the memories. Maybe data structure with `Vector` would help in this way, but I am not sure because I didn't try.

Reply:
I'm not concerned too much about it, because if traverse is vey long, we have a bigger problem because finalised block is far far away.
But I'm also not sure if this is valid `Because acc is accumulating and each iteration would create a new Seq in memory, memory would accumulate in O(n**2) way which can easily consume all the memories`. Imo with tail recursion it should be fine.

Reply:
I am not 100% sure about the cause either.Maybe it is because my implementation.
Anyway, I just want to let you and @tgrospic know about that.

The `Stream` way is copied from the old devs which is lazy -> good for memory.